README
	ASPseek v.1.2    http://www.aspseek.org/
	    Advanced Internet search engine
	Copyright (C) 2000, 2001, 2002 by SWsoft



ASPseek is a full-featured medium-to-large scale Internet search engine.
It consists of an indexing robot, a search daemon and a search front-ends
(CGI or Apache module). These programs are written in C++ using STL library.
ASPseek uses mix of SQL database and binary files for data storage.


	ASPseek features
	----------------

To learn about ASPseek features, please read aspseek(7) man page.
Here is just a brief list:

* Ability to index and search through several millions of documents
* HTTP, HTTP proxy, FTP (via proxy) protocols
* HTTP basic authorization
* HTTPS protocol
* text/html and text/plain documents
* Other document types support via external converters
* Architecture optimized for multiple sites
* Multithreaded
* Async DNS resolver
* Stopwords
* Unicode support to deal with many character sets (including CJK) at once
* Charset guesser (optional)
* Language guesser
* Robot exclusion standard (robots.txt) support
* Settings to control network bandwidth usage and Web servers load
* Real-time asynchronous indexing
* Very good relevancy of results
* Sorting results by relevance or by date
* Smart results cache
* Advanced search capabilities
* Ispell support
* Excerpts
* Grouping results by site
* Clones (mirrored documents) detection
* Spaces and subsets
* Query words highlighting in results
* Cached compressed local copy of every indexed document
* HTML templates for easy-to-customize search results


	How to use it
	-------------

Please start with reading INSTALL file there you can find detailed instructions
about installation, run-time configuration and usage of ASPseek.


	Contribution
	------------

As ASPseek is free software project, any type of contribution is
very welcome. Please send patches, bug reports, ideas etc. to
developers <aseek@sw.com.sg>

You can also subscribe to ASPseek users mailing list to ask questions,
share ideas, and communicate in general with other users and developers.
To subscribe, please send mail to majordomo@lists.asplinux.ru with the line
"subscribe aseek-users" in message body.


	Disclaimer (see COPYING for details)
	------------------------------------

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


NEWS
28 April 2007 v.1.2.15 (stable)
* Added reconnect option for MySQL5+
* Template cleaned (W3C Valid)

02 Jan 2006: v.1.2.13 (stable)
* Fixed some code
* Added default regex filter to http://xxx.xxx.xx/xx scheme
  (You can change it to ftp:// or https://) in aspseek.conf

xx Xxx 2006: v.1.2.12 (stable)
* Fixed Makefile.am
* Fixed template <> bug in new GCC compiler
* Fixed string definition in header file
* Fixed initialization of reference in header file

xx Xxx 2004: v.1.2.11 (stable)

* Fixed two bugs caused incorrect processing of some UNICODE symbols
* Fixed mistake in <!--/UdmComment--> closing tag processing
* Fixed bug that caused s.cgi to die instead of displaying error
  in case searchd is down (bug #16)
* Fixed coredump on index/searchd start if libaspseek.so can not be found
* Fixed processing of User-Agent in robots.txt (bug #19)
* Fixed problem where ^C sends all resolvers into zombied state
* Fixed s.cgi problem determining template name when Apache's AddHandler
  and Action features are used
* Fixed SGML entities in discovered HREFs not converted (bug #25)
* Fixed a problem in CUrl::ParseURL() which corrupted some HREFs
* Added rfc2396 compliant escaping of URI to CUrl::HTTPGetUrl()
* Fixed processing of charset in 'Converter' directive of aspseek.conf
* Added charset to 'Converter' directive in aspseek.conf(5) manpage
* searchd prints warning in case of unknown site is used in 'SiteWeight'
* Fixed wrong description of MaxDocSize parameter in aspseek.conf (5) man page
* Fixed compilation on systems with gcc-3.2 (such as RedHat-8.0)
* https:// now tries SSLv3 first, and can roll back to SSLv2 (bug #36)
* Fixed bug in code (triggered by gcc32 -O1) that caused first character
  of scheme string (like http://) to disappear
* Fixed bug in external converters code that prevents converter from running
  if Content-Type header returned by web server contained charset info
* Fixed words before meta robots none/noindex still indexed (bug #15)
* Added support for Last-Modified META tag
* Added --enable-max-url-length parameter to configure
* Fixed index message about stale pid file
* Fixed problem where some queries may not be logged to stat table
* Fixed incorrect calculation of chunk offset in CDelMap::IsSet()
* Fixed misc/*.m4 scripts to satisfy newer auto* tools (bug #34)
* Fixed parsing of meta http-equiv that contains quoted charset field
  (such as content='text/html; charset="KOI8-R"')
* Added --fno-strict-aliasing to CXXFLAGS that cured bad searchd looping
  when compiled with gcc3
* Fixed compilation with gcc-3.3 and libstdc++-3.3
* Doc files (like this file) are no more installed during 'make install'
* Added -F option for searchd (described in searchd(1) man page)


22 Jul 2002: v.1.2.10 (stable)

* Removed CharsetTable and LocalCharset from s.htm and manual
* Added cs parameter into s.htm search form
* Fixed bug in reverse citations merging which was introduced in 1.2.9
  (please read "Upgrade" section in INSTALL if you have used 1.2.9)
* Fixed incorrect redirects merging
* Fixed incorrect merging of reverse citation index in case when
  URL A refers to B which redirects to C and A also refers to C
  and one of references from A disappeared
* Fixed paths to s.htm in s.cgi man page
* Added CharsetAlias directives to ucharset.conf
* Fixed that ResultsPerPage parameter from s.htm overrides PS cookie (bug #12)
* Fixed password generation in aspseek-mysql-postinstall script
* Small improvements to init.d/aspseek script
* Added DBLibDir directive to aspseek.conf and searchd.conf


03 Jul 2002: v.1.2.9 (stable)

* Implemented \1 to \9 sequences in Replace command in aspseek.conf
* Fixes to work on FreeBSD
* Improved /etc/init.d/aspseek script to return proper exit code
  and shut down index while stopping
* Fixed s.cgi coredump while working with several search daemons
* Renamed SQL table 'cache' to 'rescache' to avoid collision with some DBMSes
* Fixed bug that caused -a -f file to mark all URLs for re-index (bug #8)
* Fixed bug in merging of direct citation index which resulted in incorrect
  ranks calculation
* Fixed rare memleak in searchd which occurred when one word was used more
  than once in a query
* Fixed 2 bugs in merging of reverse citations which could potentially
  lead to index coredump
* Fixed memleak in searchd when Cache was on and number of query page (np*ps)
  was greater than number of cached results
* Fixed search of words with uppercased letters with two-byte charsets
* Man pages improvements and fixes
* Fixed limiting results by range of dates (db and de) in s.cgi
* Fixed connect problem when resolvers were used on big endian architecture
* Fixed installation of docs and configs on some platforms
* "site:" (site limit) is taken into account in results cache
* Fixed rare searchd coredump which occurred when "site:" was used
* Fixed incorrect search when more than one "-site:" was used
* Fixed non-threadsafeness of query parser in searchd (it broke use
  of "-" in queries in high-load situations)
* Fixed bug of inserting newlines before ending font tag during
  cached page highlighting
* Fixed rare coredump in searchd if results cache was used
* Improved tag parsing in index to handle omitted quotes, fixes cases
  such as <A HREF="http://www.server.com"; TARGET=_new">
* Initial URL insertion (via "Server" or '-i') and URL deletion procedures
  in index now uses delmap
* Fixed problem with s.cgi when it has no suffix
* s.cgi does not prepend hostname to script_name
* Added support for HTTP method POST to s.cgi
* Number of index threads processing alive hosts is guaranteed to be no less
  than 3/4 of total number of threads
* Fixed rare coredump in searchd when cached page was queried and record
  for that page existed in "urlword" but not existed in "urlwordsNN"
* Added IncrementHopsOnRedirect and RedirectLoopLimit options to aspseek.conf
* Added stopword lists for Catalan and Hungarian languages

src/md5.cpp
/* Copyright (C) 2000, 2001, 2002 by SWsoft
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*  $Id: md5.cpp,v 1.6 2002/05/11 15:27:27 kir Exp $
    Author : Kir Kolyshkin
	Uses parts of glibc-2.1 code
*/

/* Taken by Kir from glibc-2.1 sources  and slightly modified.
 *
 * If you want to compile it with build-in test, use
 * gcc -I ../include  -DMD5_TEST md5.c -o md5
 * Compare results with that of md5sum from GNU textutils
 *
 */

#include <string.h>
#include "aspseek-cfg.h"
#include "md5.h"

/* Added by Kir */
/* digest should be char[33] */
void md5_buffer_str(char *buf, int size, char *digest)
{
	int i;
	unsigned char bin[16];
	md5_buffer(buf, (size_t)size, bin);

	if (bin)
	{
		for (i = 0; i < 16; i++)
			sprintf(&digest[i << 1], "%02x", bin[i]);
		digest[33] = '\0';
	}
	else
		digest[0] = '\0';
}

/* md5.c - Functions to compute MD5 message digest of files or memory blocks
   according to the definition of MD5 in RFC 1321 from April 1992.
   Copyright (C) 1995, 1996, 1997 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with the GNU C Library; see the file COPYING.LIB.  If not,
   write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
   Boston, MA 02111-1307, USA.  */

/* Written by Ulrich Drepper <drepper@gnu.ai.mit.edu>, 1995.  */

#include <sys/types.h>

#include <stdlib.h>
#include <string.h>


#ifdef WORDS_BIGENDIAN
#define SWAP(n) \
 (((n) << 24) | (((n) & 0xff00) << 8) | (((n) >> 8) & 0xff00) | ((n) >> 24))
#else
#define SWAP(n) (n)
#endif


/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] = { 0x80, 0 /* , 0, 0, ...  */ };


/* Initialize structure containing state of computation.
   (RFC 1321, 3.3: Step 3)  */
void
md5_init_ctx (struct md5_ctx *ctx)
{
  ctx->A = 0x67452301;
  ctx->B = 0xefcdab89;
  ctx->C = 0x98badcfe;
  ctx->D = 0x10325476;

  ctx->total[0] = ctx->total[1] = 0;
  ctx->buflen = 0;
}

/* Put result from CTX in first 16 bytes following RESBUF.  The result
   must be in little endian byte order.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_read_ctx (const struct md5_ctx* ctx, void* resbuf)
{
  ((md5_uint32 *) resbuf)[0] = SWAP (ctx->A);
  ((md5_uint32 *) resbuf)[1] = SWAP (ctx->B);
  ((md5_uint32 *) resbuf)[2] = SWAP (ctx->C);
  ((md5_uint32 *) resbuf)[3] = SWAP (ctx->D);

  return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_finish_ctx (struct md5_ctx* ctx, void* resbuf)
{
  /* Take yet unprocessed bytes into account.  */
  md5_uint32 bytes = ctx->buflen;
  size_t pad;

  /* Now count remaining bytes.  */
  ctx->total[0] += bytes;
  if (ctx->total[0] < bytes)
    ++ctx->total[1];

  pad = bytes >= 56 ? 64 + 56 - bytes : 56 - bytes;
  memcpy (&ctx->buffer[bytes], fillbuf, pad);

  /* Put the 64-bit file length in *bits* at the end of the buffer.  */
  *(md5_uint32 *) &ctx->buffer[bytes + pad] = SWAP (ctx->total[0] << 3);
  *(md5_uint32 *) &ctx->buffer[bytes + pad + 4] = SWAP ((ctx->total[1] << 3) |
							(ctx->total[0] >> 29));

  /* Process last bytes.  */
  md5_process_block (ctx->buffer, bytes + pad + 8, ctx);

  return md5_read_ctx (ctx, resbuf);
}

/* Compute MD5 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
md5_buffer (const char* buffer, size_t len, void* resblock)
{
  struct md5_ctx ctx;

  /* Initialize the computation context.  */
  md5_init_ctx (&ctx);

  /* Process whole buffer but last len % 64 bytes.  */
  md5_process_bytes (buffer, len, &ctx);

  /* Put result in desired memory area.  */
  return md5_finish_ctx (&ctx, resblock);
}


void
md5_process_bytes (const void* buffer, size_t len, struct md5_ctx* ctx)
{
  /* When we already have some bits in our internal buffer concatenate
     both inputs first.  */
  if (ctx->buflen != 0)
    {
      size_t left_over = ctx->buflen;
      size_t add = 128 - left_over > len ? len : 128 - left_over;

      memcpy (&ctx->buffer[left_over], buffer, add);
      ctx->buflen += add;

      if (left_over + add > 64)
	{
	  md5_process_block (ctx->buffer, (left_over + add) & ~63, ctx);
	  /* The regions in the following copy operation cannot overlap.  */
	  memcpy (ctx->buffer, &ctx->buffer[(left_over + add) & ~63],
		  (left_over + add) & 63);
	  ctx->buflen = (left_over + add) & 63;
	}

      buffer = (const char *) buffer + add;
      len -= add;
    }

  /* Process available complete blocks.  */
  if (len > 64)
    {
      md5_process_block (buffer, len & ~63, ctx);
      buffer = (const char *) buffer + (len & ~63);
      len &= 63;
    }

  /* Move remaining bytes in internal buffer.  */
  if (len > 0)
    {
      memcpy (ctx->buffer, buffer, len);
      ctx->buflen = len;
    }
}


/* These are the four functions used in the four steps of the MD5 algorithm
   and defined in the RFC 1321.  The first function is a little bit optimized
   (as found in Colin Plumbs public domain implementation).  */
/* #define FF(b, c, d) ((b & c) | (~b & d)) */
#define FF(b, c, d) (d ^ (b & (c ^ d)))
#define FG(b, c, d) FF (d, b, c)
#define FH(b, c, d) (b ^ c ^ d)
#define FI(b, c, d) (c ^ (b | ~d))

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.  */

void
md5_process_block (const void* buffer, size_t len, struct md5_ctx* ctx)
{
  md5_uint32 correct_words[16];
  const md5_uint32 *words = (const md5_uint32*)buffer;
  size_t nwords = len / sizeof (md5_uint32);
  const md5_uint32 *endp = words + nwords;
  md5_uint32 A = ctx->A;
  md5_uint32 B = ctx->B;
  md5_uint32 C = ctx->C;
  md5_uint32 D = ctx->D;

  /* First increment the byte count.  RFC 1321 specifies the possible
     length of the file up to 2^64 bits.  Here we only compute the
     number of bytes.  Do a double word increment.  */
  ctx->total[0] += len;
  if (ctx->total[0] < len)
    ++ctx->total[1];

  /* Process all bytes in the buffer with 64 bytes in each round of
     the loop.  */
  while (words < endp)
    {
      md5_uint32 *cwp = correct_words;
      md5_uint32 A_save = A;
      md5_uint32 B_save = B;
      md5_uint32 C_save = C;
      md5_uint32 D_save = D;

      /* First round: using the given function, the context and a constant
	 the next context is computed.  Because the algorithms processing
	 unit is a 32-bit word and it is determined to work on words in
	 little endian byte order we perhaps have to change the byte order
	 before the computation.  To reduce the work for the next steps
	 we store the swapped words in the array CORRECT_WORDS.  */

#define OP(a, b, c, d, s, T)						\
      do								\
        {								\
          md5_uint32 ttt;						\
          const void *qqq = words;					\
          memcpy((void *)&ttt,qqq,sizeof(ttt));				\
	  a += FF (b, c, d) + (*cwp++ = SWAP (ttt)) + T;		\
	  ++words;							\
	  CYCLIC (a, s);						\
	  a += b;							\
        }								\
      while (0)

      /* It is unfortunate that C does not provide an operator for
	 cyclic rotation.  Hope the C compiler is smart enough.  */
#define CYCLIC(w, s) (w = (w << s) | (w >> (32 - s)))

      /* Before we start, one word to the strange constants.
	 They are defined in RFC 1321 as

	 T[i] = (int) (4294967296.0 * fabs (sin (i))), i=1..64
       */

      /* Round 1.  */
      OP (A, B, C, D,  7, 0xd76aa478);
      OP (D, A, B, C, 12, 0xe8c7b756);
      OP (C, D, A, B, 17, 0x242070db);
      OP (B, C, D, A, 22, 0xc1bdceee);
      OP (A, B, C, D,  7, 0xf57c0faf);
      OP (D, A, B, C, 12, 0x4787c62a);
      OP (C, D, A, B, 17, 0xa8304613);
      OP (B, C, D, A, 22, 0xfd469501);
      OP (A, B, C, D,  7, 0x698098d8);
      OP (D, A, B, C, 12, 0x8b44f7af);
      OP (C, D, A, B, 17, 0xffff5bb1);
      OP (B, C, D, A, 22, 0x895cd7be);
      OP (A, B, C, D,  7, 0x6b901122);
      OP (D, A, B, C, 12, 0xfd987193);
      OP (C, D, A, B, 17, 0xa679438e);
      OP (B, C, D, A, 22, 0x49b40821);

      /* For the second to fourth round we have the possibly swapped words
	 in CORRECT_WORDS.  Redefine the macro to take an additional first
	 argument specifying the function to use.  */
#undef OP
#define OP(f, a, b, c, d, k, s, T)					\
      do 								\
	{								\
	  a += f (b, c, d) + correct_words[k] + T;			\
	  CYCLIC (a, s);						\
	  a += b;							\
	}								\
      while (0)

      /* Round 2.  */
      OP (FG, A, B, C, D,  1,  5, 0xf61e2562);
      OP (FG, D, A, B, C,  6,  9, 0xc040b340);
      OP (FG, C, D, A, B, 11, 14, 0x265e5a51);
      OP (FG, B, C, D, A,  0, 20, 0xe9b6c7aa);
      OP (FG, A, B, C, D,  5,  5, 0xd62f105d);
      OP (FG, D, A, B, C, 10,  9, 0x02441453);
      OP (FG, C, D, A, B, 15, 14, 0xd8a1e681);
      OP (FG, B, C, D, A,  4, 20, 0xe7d3fbc8);
      OP (FG, A, B, C, D,  9,  5, 0x21e1cde6);
      OP (FG, D, A, B, C, 14,  9, 0xc33707d6);
      OP (FG, C, D, A, B,  3, 14, 0xf4d50d87);
      OP (FG, B, C, D, A,  8, 20, 0x455a14ed);
      OP (FG, A, B, C, D, 13,  5, 0xa9e3e905);
      OP (FG, D, A, B, C,  2,  9, 0xfcefa3f8);
      OP (FG, C, D, A, B,  7, 14, 0x676f02d9);
      OP (FG, B, C, D, A, 12, 20, 0x8d2a4c8a);

      /* Round 3.  */
      OP (FH, A, B, C, D,  5,  4, 0xfffa3942);
      OP (FH, D, A, B, C,  8, 11, 0x8771f681);
      OP (FH, C, D, A, B, 11, 16, 0x6d9d6122);
      OP (FH, B, C, D, A, 14, 23, 0xfde5380c);
      OP (FH, A, B, C, D,  1,  4, 0xa4beea44);
      OP (FH, D, A, B, C,  4, 11, 0x4bdecfa9);
      OP (FH, C, D, A, B,  7, 16, 0xf6bb4b60);
      OP (FH, B, C, D, A, 10, 23, 0xbebfbc70);
      OP (FH, A, B, C, D, 13,  4, 0x289b7ec6);
      OP (FH, D, A, B, C,  0, 11, 0xeaa127fa);
      OP (FH, C, D, A, B,  3, 16, 0xd4ef3085);
      OP (FH, B, C, D, A,  6, 23, 0x04881d05);
      OP (FH, A, B, C, D,  9,  4, 0xd9d4d039);
      OP (FH, D, A, B, C, 12, 11, 0xe6db99e5);
      OP (FH, C, D, A, B, 15, 16, 0x1fa27cf8);
      OP (FH, B, C, D, A,  2, 23, 0xc4ac5665);

      /* Round 4.  */
      OP (FI, A, B, C, D,  0,  6, 0xf4292244);
      OP (FI, D, A, B, C,  7, 10, 0x432aff97);
      OP (FI, C, D, A, B, 14, 15, 0xab9423a7);
      OP (FI, B, C, D, A,  5, 21, 0xfc93a039);
      OP (FI, A, B, C, D, 12,  6, 0x655b59c3);
      OP (FI, D, A, B, C,  3, 10, 0x8f0ccc92);
      OP (FI, C, D, A, B, 10, 15, 0xffeff47d);
      OP (FI, B, C, D, A,  1, 21, 0x85845dd1);
      OP (FI, A, B, C, D,  8,  6, 0x6fa87e4f);
      OP (FI, D, A, B, C, 15, 10, 0xfe2ce6e0);
      OP (FI, C, D, A, B,  6, 15, 0xa3014314);
      OP (FI, B, C, D, A, 13, 21, 0x4e0811a1);
      OP (FI, A, B, C, D,  4,  6, 0xf7537e82);
      OP (FI, D, A, B, C, 11, 10, 0xbd3af235);
      OP (FI, C, D, A, B,  2, 15, 0x2ad7d2bb);
      OP (FI, B, C, D, A,  9, 21, 0xeb86d391);

      /* Add the starting values of the context.  */
      A += A_save;
      B += B_save;
      C += C_save;
      D += D_save;
    }

  /* Put checksum in context given as argument.  */
  ctx->A = A;
  ctx->B = B;
  ctx->C = C;
  ctx->D = D;
}
